content/blog/2024-01-09-xilem-2024.md

```diff
@@ -0,0 +1,80 @@
+
+### Declarative UI patterns
+
+The trend in UI programming has been overwhelmingly away from a soup of mutable objects with interlinked references and toward various declarative or reactive patterns. That’s a good thing for Rust UI, because the former is very un-ergonomic.
+
+In the Javascript world, there is a Cambrian explosion of reactive techniques. React is by far the most popular, but there is a strong sentiment it’s possible to be better, particularly on performance, but also robustness, as the logic for deciding [when to re-render](https://www.joshwcomeau.com/react/why-react-re-renders/) is fickle, and can [miss re-renders](https://arxiv.org/abs/2310.04368). Elm deserves special mention, as its avoidance of global shared mutability makes it particularly easy to adapt to Rust, so it serves as the reactive core of the iced, relm, and vizia UI toolkits, as well as yew on the WASM side. But there is also Ember Octane (based on Adapton), a number of fine-grained reactivity approaches such as SolidJS, and many, many more.
+
+We’re seeing much of that diversity mirrored into the Rust world. On the WASM side, we see Dioxus as a fairly faithful adaptation of React, leptos (also sycamore and tachys) adhering fairly closely to SolidJS, yew as an adaption of elm, and [more](https://github.com/flosse/rust-web-framework-comparison).
+
+A reactive layer such as Dioxus or leptos can be compiled to WASM and run in a browser to generate DOM nodes. More recently, we’re seeing that it can be decoupled from that environment, and a reactive core can drive a widget tree. Examples include [Freya](https://github.com/marc2332/freya) which uses Dioxus, and [floem](https://github.com/lapce/floem) which implements fine-grained reactivity inspired by leptos.
+
+Xilem doesn’t directly correspond with any Javascript framework; it relies on static typing and is actually closer to SwiftUI than anything web based.
+
+A major development with xilem is its factoring into xilem_core, supporting both xilem_web for DOM, and our own bespoke widget set.
+
+One of the central questions in Rust UI going forward is: which approach to declarative UI is best? It’s possible a clear winner will emerge, or perhaps there will be a different answer depending on the use case, or it might just come down to a matter of personal style, with a number of viable contenders (arguably that’s where Javascript is). We’re taking a two pronged approach. First, we’re moving forward with Xilem. There’s a specific reason - Xilem is designed to support multithreaded rendering easily, while many of the other frameworks rely on thread local storage for access to a store of observables, an approach with high impedance mismatch to multithreading. Since performance is a goal, we didn’t want to give that up by adopting another framework. I also think Xilem is pretty good, but that remains to be seen.
+
```

Comment:
I'll be so bold to comment here. I feel like this transition is still pretty rough. The last half sentence of the last paragraph started already with a `but`, as does the next paragraph. The shift from UI paradigms to the current state of xilem feels a bit abrupt.